common/quotas/global/collection/collection.go

```diff
@@ -0,0 +1,251 @@
+*/
+package collection
+
+import (
+	"context"
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/uber/cadence/common/clock"
+	"github.com/uber/cadence/common/dynamicconfig"
+	"github.com/uber/cadence/common/log"
+	"github.com/uber/cadence/common/log/tag"
+	"github.com/uber/cadence/common/metrics"
+	"github.com/uber/cadence/common/quotas"
+	"github.com/uber/cadence/common/quotas/global/collection/internal"
+)
+
+type (
+	Collection struct {
+		updateRate dynamicconfig.DurationPropertyFn
+
+		logger log.Logger
+		scope  metrics.Scope
+
+		limits *internal.AtomicMap[string, *internal.FallbackLimiter]
+
+		ctx       context.Context // context used for background operations, canceled when stopping
+		ctxCancel func()
+		stopped   chan struct{} // closed when stopping is complete
+
+		// now exists largely for tests, elsewhere it is always time.Now
+		timesource clock.TimeSource
+	}
+
+	// Limiter is a simplified quotas.Limiter API, covering the only global-ratelimiter-used API: Allow.
+	//
+	// This is largely because Reserve() and Wait() are difficult to monitor for usage-tracking purposes.
+	// If converted to the wrapper, Reserve can probably be added pretty easily.
+	Limiter interface {
+		Allow() bool
+	}
+
+	// calls is a small temporary pair[T] container
+	calls struct {
+		allowed, rejected int
+	}
+)
+
+func New(
+	fallback quotas.LimiterFactory,
+	updateRate dynamicconfig.DurationPropertyFn,
+	logger log.Logger,
+	met metrics.Client,
+) *Collection {
+	contents := internal.NewAtomicMap(func(key string) *internal.FallbackLimiter {
+		return internal.NewFallbackLimiter(fallback.GetLimiter(key))
+	})
+	ctx, cancel := context.WithCancel(context.Background())
+	c := &Collection{
+		limits:     contents,
+		updateRate: updateRate,
+
+		logger: logger.WithTags(tag.ComponentGlobalRatelimiter),
+		scope:  met.Scope(metrics.FrontendGlobalRatelimiter),
+
+		ctx:       ctx,
+		ctxCancel: cancel,
+		stopped:   make(chan struct{}),
+
+		// override externally in tests
+		timesource: clock.NewRealTimeSource(),
+	}
+	return c
+}
+
+func (c *Collection) TestOverrides(t *testing.T, timesource clock.TimeSource) {
+	t.Helper()
+	c.timesource = timesource
+}
+
+// OnStart follows fx's OnStart hook semantics.
+func (c *Collection) OnStart(ctx context.Context) error {
+	go func() {
+		defer func() {
+			// bad but not worth crashing the process
+			log.CapturePanic(recover(), c.logger, nil)
+		}()
+		defer func() {
+			close(c.stopped)
+		}()
+		c.backgroundUpdateLoop()
+	}()
+
+	return nil
+}
+
+// OnStop follows fx's OnStop hook semantics.
+func (c *Collection) OnStop(ctx context.Context) error {
+	c.ctxCancel()
+	// context timeout is for everything to shut down, but this one should be fast.
+	// don't wait very long.
+	ctx, cancel := context.WithTimeout(ctx, 100*time.Millisecond)
+	defer cancel()
+
+	select {
+	case <-ctx.Done():
+		return fmt.Errorf("*Collection failed to stop, context canceled: %w", ctx.Err())
```

Comment:
nit: I think it is very unexpected to see pointer in error-log.

Reply:
huh.  yeah, I think that got caught in a refactor 🤔 
changing 👍